torchmultimodal/modules/optimizers/anyprecision.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import Any, Dict, Iterable, Tuple, Union
+
+import torch
+from torch.optim.optimizer import Optimizer
+from typing_extensions import TypeAlias
+
+params_t: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
+
+
+class AnyPrecisionAdamW(Optimizer):
+    def __init__(
+        self,
+        params: params_t,
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 0.0,
+        use_kahan_summation: bool = False,
+        momentum_dtype: torch.dtype = torch.float32,
+        variance_dtype: torch.dtype = torch.bfloat16,
+        compensation_buffer_dtype: torch.dtype = torch.bfloat16,
```

Comment:
only A100 and more recent support bf16, so should we choose defaults that are more hardware agnostic?

Reply:
hmm - BF16 is pretty mainstream now.  
The Ampere architecture supports native BF16, and has been around since 2020 and that includes datacenter A2, A10, A100 as well as consumer 2050 (mobile), 3050 - 3090...4090 range.   
Hopper of course supports it as well though that is not common atm.
For the compensation buffer and variance though, FP16 should be fine as we don't really need the full range of BF16 and the extra precision of FP16 could be handy.

Ok will switch over to FP16 as the default then, but will need to update the unit tests to match.

Reply:
well nvm - we need to stay with BF16.
In running the unit tests using FP16 and inspecting the compensation buffer as fp16, started seeing 'nans' which I did not expect but will defeat the purpose of the compensation buffer.
Anyway, will stick with BF16 and based on how mainstream BF16 is at this point don't think it should be an issue.